src/Common/Commands.Common.Authentication/Authentication/ProtectedFileTokenCache.cs

```diff
@@ -113,45 +96,70 @@ public override void Clear()
         // Reload the cache from the persistent store in case it changed since the last access.
         void BeforeAccessNotification(TokenCacheNotificationArgs args)
         {
+            ReadFileIntoCache();
+        }
+
+        // Triggered right after ADAL accessed the cache.
+        void AfterAccessNotification(TokenCacheNotificationArgs args)
+        {
+            // if the access operation resulted in a cache update
+            EnsureStateSaved();
+        }
+
+        void EnsureStateSaved()
+        {
+            if (HasStateChanged)
```

Comment:
this isn't quite the same logic as before - we only want to write the cache once after a state change, so we want to write and clear the dirty bit in an atomic operation and keep other threads from doing so.

We can do this with a pattern like:

```C#
if (HasStateChanged)
{
   WriteCacheInfoFile()
}
```

and inside WriteCacheInfoFile:

``` C#
lock(lockFile)
{
  if (HasStateChanged)
  {  
   store.WriteFile(...);
   HasStateChanged =  false;
  }
}
```

or an equivalent.